### ISSUE TITLE
Builder append mode: Error during upload of: index.docker.io/.../fairing-job:2B7FAA2D: BrokenPipeError

### ISSUE THREAD

**Issue by marcindulak** [kubeflow/fairing#461]
/kind bug

**What steps did you take and what happened:**

Fairing fails to push the image to a dockerhub repository (private or public, created manually in advance on dockerhub).

I'm running jupyter image `gcr.io/kubeflow-images-public/tensorflow-1.14.0-notebook-cpu:v0.7.0` as the jupyter server, started with 1 core and 4GB RAM.

Providing the docker credentials by creating `.docker/config.json` inside of the jupyter server container, with the contents generated by `docker login` on a host with docker installed.

Fairing installed from a notebook cell with

```
!pip install --upgrade --user kubeflow-fairing
```

```
!pip show kubeflow-fairing
Name: kubeflow-fairing
Version: 0.7.1
Summary: Kubeflow Fairing Python SDK.
Home-page: https://github.com/kubeflow/fairing
Author: Kubeflow Authors
Author-email: hejinchi@cn.ibm.com
License: Apache License Version 2.0
Location: /home/jovyan/.local/lib/python3.6/site-packages
Requires: numpy, kubeflow-tfjob, python-dateutil, six, cloudpickle, kubernetes, boto3, setuptools, azure, kfserving, notebook, docker, urllib3, google-auth, retrying, requests, google-api-python-client, future, oauth2client, kubeflow-pytorchjob, google-cloud-logging, httplib2, google-cloud-storage, tornado
Required-by:  
```

The modified https://github.com/kubeflow/fairing/blob/7e4c23c47a6b0544a722d04e1f9910c2f283a4f6/examples/simple/main.py test job

```
import os
import tensorflow as tf

from kubeflow import fairing
DOCKER_REGISTRY = 'index.docker.io/...'  # where ... is a dockerhub username
fairing.config.set_builder('append',
                           base_image='tensorflow/tensorflow:latest-py3',
                           registry=DOCKER_REGISTRY, push=True)
fairing.config.set_deployer('job')

def train():
    hostname = tf.constant(os.environ['HOSTNAME'])
    sess = tf.Session()
    print('Hostname: ', sess.run(hostname).decode('utf-8'))

if __name__ == '__main__':
    remote_train = fairing.config.fn(train)
    remote_train()
```

fails with

```
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
WARNING: Logging before flag parsing goes to stderr.
I0220 01:13:09.378802 140433427404608 config.py:125] Using preprocessor: <kubeflow.fairing.preprocessors.function.FunctionPreProcessor object at 0x7fb8a45404e0>
I0220 01:13:09.379966 140433427404608 config.py:127] Using builder: <kubeflow.fairing.builders.append.append.AppendBuilder object at 0x7fb8a4540e80>
I0220 01:13:09.381677 140433427404608 config.py:129] Using deployer: <kubeflow.fairing.deployers.job.job.Job object at 0x7fb900fbcd68>
W0220 01:13:09.382750 140433427404608 append.py:50] Building image using Append builder...
I0220 01:13:09.384421 140433427404608 base.py:107] Creating docker context: /tmp/fairing_context_9rdlswve
W0220 01:13:09.385552 140433427404608 base.py:94] /home/jovyan/.local/lib/python3.6/site-packages/kubeflow/fairing/__init__.py already exists in Fairing context, skipping...
I0220 01:13:09.399490 140433427404608 docker_creds_.py:234] Loading Docker credentials for repository 'tensorflow/tensorflow:2.0.1-py3'
W0220 01:13:11.413914 140433427404608 append.py:54] Image successfully built in 2.0301002760024858s.
W0220 01:13:11.414849 140433427404608 append.py:94] Pushing image index.docker.io/.../fairing-job:4720854A...
I0220 01:13:11.415671 140433427404608 docker_creds_.py:234] Loading Docker credentials for repository 'index.docker.io/.../fairing-job:4720854A'
W0220 01:13:11.673852 140433427404608 append.py:81] Uploading index.docker.io/.../fairing-job:4720854A
I0220 01:13:12.002556 140429044655872 docker_session_.py:280] Layer sha256:651c9d2d6c4f37c56a221259e033e7e2353b698139c2ff950623ca28d64a9837 exists, skipping
...
I0220 01:13:14.513353 140428985644800 docker_session_.py:284] Layer sha256:5365cd9485e2b9e501ccc51efebcd85cb975384359a2b067cffe6c81ccc6ac91 pushed.
E0220 01:15:04.274829 140433427404608 docker_session_.py:332] Error during upload of: index.docker.io/.../fairing-job:4720854A
---------------------------------------------------------------------------
BrokenPipeError                           Traceback (most recent call last)
<ipython-input-31-55b6073a213c> in <module>
     15 if __name__ == '__main__':
     16     remote_train = fairing.config.fn(train)
---> 17     remote_train()

~/.local/lib/python3.6/site-packages/kubeflow/fairing/config.py in ret_fn()
    152             """ return the function"""
    153             self.set_preprocessor('function', function_obj=fn)
--> 154             self.run()
    155         return ret_fn
    156 

~/.local/lib/python3.6/site-packages/kubeflow/fairing/config.py in run(self)
    129         logging.info("Using deployer: %s", deployer)
    130 
--> 131         builder.build()
    132         pod_spec = builder.generate_pod_spec()
    133         deployer.deploy(pod_spec)

~/.local/lib/python3.6/site-packages/kubeflow/fairing/builders/append/append.py in build(self)
     56             self.full_image_name(self.context_hash), strict=False)
     57         if self.push:
---> 58             self.timed_push(transport, src, new_img, dst)
     59         else:
     60             # TODO(r2d4):

~/.local/lib/python3.6/site-packages/kubeflow/fairing/builders/append/append.py in timed_push(self, transport, src, img, dst)
     94         logger.warning("Pushing image {}...".format(self.image_tag))
     95         start = timer()
---> 96         self._push(transport, src, img, dst)
     97         end = timer()
     98         logger.warning(

~/.local/lib/python3.6/site-packages/kubeflow/fairing/builders/append/append.py in _push(self, transport, src, img, dst)
     80                                  mount=[src.as_repository()]) as session:
     81             logger.warning("Uploading {}".format(self.image_tag))
---> 82             session.upload(img)
     83         os.remove(self.context_file)
     84 

~/.local/lib/python3.6/site-packages/containerregistry/client/v2_2/docker_session_.py in upload(self, image, use_digest)
    319         }
    320         for future in concurrent.futures.as_completed(future_to_params):
--> 321           future.result()
    322 
    323     # This should complete the upload by uploading the manifest.

/usr/lib/python3.6/concurrent/futures/_base.py in result(self, timeout)
    423                 raise CancelledError()
    424             elif self._state == FINISHED:
--> 425                 return self.__get_result()
    426 
    427             self._condition.wait(timeout)

/usr/lib/python3.6/concurrent/futures/_base.py in __get_result(self)
    382     def __get_result(self):
    383         if self._exception:
--> 384             raise self._exception
    385         else:
    386             return self._result

/usr/lib/python3.6/concurrent/futures/thread.py in run(self)
     54 
     55         try:
---> 56             result = self.fn(*self.args, **self.kwargs)
     57         except BaseException as exc:
     58             self.future.set_exception(exc)

~/.local/lib/python3.6/site-packages/containerregistry/client/v2_2/docker_session_.py in _upload_one(self, image, digest)
    281       return
    282 
--> 283     self._put_blob(image, digest)
    284     logging.info('Layer %s pushed.', digest)
    285 

~/.local/lib/python3.6/site-packages/containerregistry/client/v2_2/docker_session_.py in _put_blob(self, image, digest)
    204     # specified in the "mount" parameter. This does a fast copy from a
    205     # repository that is known to contain this blob and skips the upload.
--> 206     self._patch_upload(image, digest)
    207 
    208   def _remote_tag_digest(

~/.local/lib/python3.6/site-packages/containerregistry/client/v2_2/docker_session_.py in _patch_upload(self, image, digest)
    170         accepted_codes=[
    171             six.moves.http_client.NO_CONTENT, six.moves.http_client.ACCEPTED,
--> 172             six.moves.http_client.CREATED
    173         ])
    174 

~/.local/lib/python3.6/site-packages/containerregistry/client/v2_2/docker_http_.py in Request(self, url, accepted_codes, method, body, content_type, accepted_mimes)
    384 
    385       resp, content = self._transport.request(
--> 386           url, method, body=body, headers=headers)
    387 
    388       if resp.status != six.moves.http_client.UNAUTHORIZED:

~/.local/lib/python3.6/site-packages/containerregistry/transport/transport_pool_.py in request(self, *args, **kwargs)
     59     transport = self._get_transport()
     60     try:
---> 61       return transport.request(*args, **kwargs)
     62     finally:
     63       self._return_transport(transport)

/usr/local/lib/python3.6/dist-packages/httplib2/__init__.py in request(self, uri, method, body, headers, redirections, connection_type)
   1955                         headers,
   1956                         redirections,
-> 1957                         cachekey,
   1958                     )
   1959         except Exception as e:

/usr/local/lib/python3.6/dist-packages/httplib2/__init__.py in _request(self, conn, host, absolute_uri, request_uri, method, body, headers, redirections, cachekey)
   1620 
   1621         (response, content) = self._conn_request(
-> 1622             conn, request_uri, method, body, headers
   1623         )
   1624 

/usr/local/lib/python3.6/dist-packages/httplib2/__init__.py in _conn_request(self, conn, request_uri, method, body, headers)
   1527                 if conn.sock is None:
   1528                     conn.connect()
-> 1529                 conn.request(method, request_uri, body, headers)
   1530             except socket.timeout:
   1531                 conn.close()

/usr/lib/python3.6/http/client.py in request(self, method, url, body, headers, encode_chunked)
   1237                 encode_chunked=False):
   1238         """Send a complete request to the server."""
-> 1239         self._send_request(method, url, body, headers, encode_chunked)
   1240 
   1241     def _send_request(self, method, url, body, headers, encode_chunked):

/usr/lib/python3.6/http/client.py in _send_request(self, method, url, body, headers, encode_chunked)
   1283             # default charset of iso-8859-1.
   1284             body = _encode(body, 'body')
-> 1285         self.endheaders(body, encode_chunked=encode_chunked)
   1286 
   1287     def getresponse(self):

/usr/lib/python3.6/http/client.py in endheaders(self, message_body, encode_chunked)
   1232         else:
   1233             raise CannotSendHeader()
-> 1234         self._send_output(message_body, encode_chunked=encode_chunked)
   1235 
   1236     def request(self, method, url, body=None, headers={}, *,

/usr/lib/python3.6/http/client.py in _send_output(self, message_body, encode_chunked)
   1063                     chunk = f'{len(chunk):X}\r\n'.encode('ascii') + chunk \
   1064                         + b'\r\n'
-> 1065                 self.send(chunk)
   1066 
   1067             if encode_chunked and self._http_vsn == 11:

/usr/lib/python3.6/http/client.py in send(self, data)
    984             return
    985         try:
--> 986             self.sock.sendall(data)
    987         except TypeError:
    988             if isinstance(data, collections.Iterable):

/usr/lib/python3.6/ssl.py in sendall(self, data, flags)
    973                 amount = len(byte_view)
    974                 while count < amount:
--> 975                     v = self.send(byte_view[count:])
    976                     count += v
    977         else:

/usr/lib/python3.6/ssl.py in send(self, data, flags)
    942                     "non-zero flags not allowed in calls to send() on %s" %
    943                     self.__class__)
--> 944             return self._sslobj.write(data)
    945         else:
    946             return socket.send(self, data, flags)

/usr/lib/python3.6/ssl.py in write(self, data)
    640         The 'data' argument must support the buffer interface.
    641         """
--> 642         return self._sslobj.write(data)
    643 
    644     def getpeercert(self, binary_form=False):

BrokenPipeError: [Errno 32] Broken pipe
```

Possibly related issues https://github.com/kubeflow/fairing/issues/380 https://github.com/kubeflow/fairing/issues/340 (or maybe just a bad network when pushing a large image)

**What did you expect to happen:**

Upload the image to the dockerhub private repository.
Trying to workaround https://github.com/kubeflow/fairing/issues/460

**Environment:**

- Fairing version: (use `python -c "import kubeflow.fairing; print(kubeflow.fairing.__version__)"`): 0.7.0 (pip show says 0.7.1)
- Kubeflow version: (version number can be found at the bottom left corner of the Kubeflow dashboard): 0.7.0
- Minikube version: using kind https://hub.docker.com/layers/kindest/node/v1.15.7/images/sha256-e2df133f80ef633c53c0200114fce2ed5e1f6947477dbc83261a6a921169488d?context=explore
- Kubernetes version: (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.7", GitCommit:"6c143d35bb11d74970e7bc0b6c45b6bfdffc0bd4", GitTreeState:"clean", BuildDate:"2019-12-11T12:42:56Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.7", GitCommit:"6c143d35bb11d74970e7bc0b6c45b6bfdffc0bd4", GitTreeState:"clean", BuildDate:"2020-01-14T00:28:37Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
- OS (e.g. from `/etc/os-release`):
```
cat /etc/*release | grep PRETTY
PRETTY_NAME="Ubuntu 19.10"
```

**Comment by marcindulak** [comment]
Additional experiment using a gcr.io registry also failed with the same error.

I'm using the following trick to inject the gcp credentials into the running notebook

1. exec into the jupyter pod `kubectl -n anonymous exec -it production-notebook-0 -- bash` and docker login to the gcr

```
wget https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-281.0.0-linux-x86_64.tar.gz
tar zxf google-cloud-sdk-281.0.0-linux-x86_64.tar.gz
export PATH=google-cloud-sdk/bin:$PATH
gcloud init
gcloud auth configure-docker
```

2. then setup the environment and variables from a jupyter notebook cell

```
import os
import glob
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = (
    list(map(os.path.abspath, glob.glob('.config/gcloud/legacy_credentials/*/adc.json')))[0])
os.environ['GOOGLE_CLOUD_PROJECT'] = 'my-project-197109'
```

```
GCP_PROJECT = fairing.cloud.gcp.guess_project_name()
DOCKER_REGISTRY = 'gcr.io/{}/fairing-job'.format(GCP_PROJECT)
PY_VERSION = ".".join([str(x) for x in sys.version_info[0:3]])
BASE_IMAGE = 'python:{}'.format(PY_VERSION)
BASE_IMAGE = 'gcr.io/kubeflow-images-public/tensorflow-1.14.0-notebook-cpu:v0.7.0'
```

**Comment by marcindulak** [comment]
Update: after several attempts (re-running the `remote_train()`) succeeds, but then the `fairing-job` fails to pull a private image which is https://github.com/kubeflow/fairing/issues/380

Closing, but it would be nice to have a possibility to use self-hosted registries (insecure or with a self-signed certificate) https://github.com/kubeflow/fairing/issues/460

**Comment by marcindulak** [comment]
Let me reopen until I can test pushing from a better network.

It is interesting to see that every execution of `remote_train()` keeps adding resident memory

```
top - 00:57:39 up 23:32,  0 users,  load average: 5.47, 5.98, 5.74
Tasks:   6 total,   2 running,   4 sleeping,   0 stopped,   0 zombie
%Cpu(s): 22.1 us,  8.2 sy,  0.0 ni, 68.5 id,  0.9 wa,  0.0 hi,  0.3 si,  0.0 st
KiB Mem : 32797464 total,  6001540 free, 11946336 used, 14849588 buff/cache
KiB Swap:   999420 total,   821468 free,   177952 used. 19405404 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 2639 jovyan    20   0 2367904 252596 105444 R  30.3  0.8   0:03.49 python3
    9 jovyan    20   0  355480  69724  17944 S   0.3  0.2   0:59.45 jupyter-noteboo
 2133 jovyan    20   0   20180   3944   3480 S   0.0  0.0   0:00.16 bash
 2630 jovyan    20   0   38280   3556   3084 R   0.0  0.0   0:01.22 top
    8 jovyan    20   0    4628    828    760 S   0.0  0.0   0:00.00 sh
    1 jovyan    20   0    4520    768    720 S   0.0  0.0   0:01.27 tini
```

```
top - 01:48:43 up 1 day, 23 min,  0 users,  load average: 9.61, 8.01, 6.69
Tasks:   6 total,   1 running,   5 sleeping,   0 stopped,   0 zombie
%Cpu(s): 17.1 us, 16.1 sy,  0.3 ni, 63.7 id,  0.6 wa,  0.0 hi,  2.2 si,  0.0 st
KiB Mem : 32797464 total,  4165340 free, 14090860 used, 14541264 buff/cache
KiB Swap:   999420 total,   821980 free,   177440 used. 17198484 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND 
 8464 jovyan    20   0 6117724 2.360g 121844 S   8.0  7.5   3:48.68 python3
    9 jovyan    20   0  355736  70008  17944 S   0.0  0.2   1:04.10 jupyter-noteboo
 2133 jovyan    20   0   20180   3944   3480 S   0.0  0.0   0:00.16 bash
 2630 jovyan    20   0   38280   3556   3084 R   0.3  0.0   0:03.04 top
    8 jovyan    20   0    4628    828    760 S   0.0  0.0   0:00.00 sh
    1 jovyan    20   0    4520    768    720 S   0.0  0.0   0:01.39 tini
```

```
top - 01:54:11 up 1 day, 28 min,  0 users,  load average: 9.20, 7.22, 6.68
Tasks:   6 total,   1 running,   5 sleeping,   0 stopped,   0 zombie
%Cpu(s): 20.7 us, 13.7 sy,  0.0 ni, 62.0 id,  0.7 wa,  0.0 hi,  2.9 si,  0.0 st
KiB Mem : 32797464 total,  2577852 free, 15594080 used, 14625532 buff/cache
KiB Swap:   999420 total,   822236 free,   177184 used. 15657520 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND 
 8464 jovyan    20   0 7454188 3.660g 121844 S  12.3 11.7   4:30.52 python3 
    9 jovyan    20   0  355736  70008  17944 S   0.0  0.2   1:04.73 jupyter-noteboo 
 2133 jovyan    20   0   20180   3944   3480 S   0.0  0.0   0:00.16 bash   
 2630 jovyan    20   0   38280   3556   3084 R   0.0  0.0   0:03.23 top   
    8 jovyan    20   0    4628    828    760 S   0.0  0.0   0:00.00 sh 
    1 jovyan    20   0    4520    768    720 S   0.0  0.0   0:01.39 tini  
```

**Comment by jinchihe** [comment]
```
E0220 01:15:04.274829 140433427404608 docker_session_.py:332] Error during upload of: index.docker.io/.../fairing-job:4720854A
```

Would you try to update the `DOCKER_REGISTRY ` to your docker username? such as `index.docker.io/your_docker_id`? seems that's caused uploading error. Also you can try push the image manually to see if the error. Thanks.

**Comment by anupash147** [comment]
I have a similar error when i am trying to append function from an existing docker registry which is completed using docker secrets attached to the notebook container, but the job submitted failed to have the same docker-secret to be added to even run. Solution would be to add pod_spec_mutator
